### PR TITLE
TDP: Adds student financial capabilities surveys

### PR DESCRIPTION
**Pull request by mrclay**
This adds 3 web-based surveys about financial knowledge for grade school children ([PDF versions are here](https://www.consumerfinance.gov/consumer-tools/educator-tools/youth-financial-education/assess/)), as well as a system to embed resulting scores in a tamper-proof URL which can be shared with a teacher.

The shared URL also contains the student's initials obfuscated in a fragment (see `encode-name.js`) so that this info is never sent to the CFPB server nor captured by analytics. All survey temporary storage is based on client cookies, not DB/file sessions.

---

## Additions

- Mostly static survey intro pages for 3 grade ranges
- Multi-page survey system based on the Django `formtools` FormWizard component
- A python system to create a logical survey with pages of questions with all content loaded from hardcoded CSV
- Survey object methods to generate a custom form class for a single survey page
- A few Django form subclasses to customize markup within the dynamically created forms
- A python system to supply dynamic results (also from CSV) to a template based on user scores
- JS scripts to reduce data loss if a student refreshes/returns to a previous page
- JS scripts to encode/decode the user's initials, copy the shared URL to clipboard, and track progress
- JS scripts to reduce chances of a user accidentally leaking info to a later user

## How to test this PR

1. Navigate to a grade level page (see URLs below)
2. Enter the survey
3. Complete the survey
4. Choose to share the URL, enter initials, and copy the URL
5. The shared version requires no state and can be opened in an incognito window.

- https://cfgov-pr-6511.demo.cfpb.gov/consumer-tools/educator-tools/youth-financial-education/survey/3-5/
- https://cfgov-pr-6511.demo.cfpb.gov/consumer-tools/educator-tools/youth-financial-education/survey/6-8/
- https://cfgov-pr-6511.demo.cfpb.gov/consumer-tools/educator-tools/youth-financial-education/survey/9-12/

## Screenshots

Screenshot from 2021-08-20 14-23-06
Screenshot from 2021-08-20 14-23-34
Screenshot from 2021-08-20 14-24-07
Screenshot from 2021-08-20 14-24-25

## Notes and todos

- Backend tests will be added

## Checklist

- [x] PR has an informative and human-readable title
  - PR titles are used to generate the change log in [releases](../../releases); good ones make that easier to scan.
  - Consider prefixing, e.g., "Mega Menu: fix layout bug", or "Docs: Update Docker installation instructions".
- [x] Changes are limited to a single goal (no scope creep)
- [x] Code follows the standards laid out in the [CFPB development guidelines](https://github.com/cfpb/development)
- [x] Future todos are captured in comments and/or tickets
- [ ] Project documentation has been updated, potentially one or more of:
  - [This repo’s docs](https://cfpb.github.io/consumerfinance.gov/) (edit the files in the `/docs` folder) – for basic, close-to-the-code docs on working with this repo
  - CFGOV/platform wiki on GHE – for internal CFPB developer guidance
  - CFPB/hubcap wiki on GHE – for internal CFPB design and content guidance

### Front-end testing

...

#### Browser testing

Aug 20 note: We're beginning more extensive testing this afternoon. We know that the use of the OS-default radio buttons results in, .e.g. red buttons in Firefox, however we have a plan to re-align the Django form markup with CFPB standards as soon as possible.

Visually tested in the following supported browsers:
- [ ] Firefox
- [x] Chrome
- [x] Safari
- [ ] Edge 18 (the last Edge prior to it switching to Chromium)
- [ ] Internet Explorer 11 and 8 (via emulation in 11's dev tools)
- [ ] Safari on iOS
- [ ] Chrome on Android

#### Accessibility

Aug 0 note: I found a WAVE warning about duplicate radio button labels because of a Django form markup bug. We're hoping to address when re revisit the radio markup.

- [x] Keyboard friendly (navigable with tab, space, enter, arrow keys, etc.)
- [x] Screen reader friendly
- [ ] Does not introduce new errors or warnings in [WAVE](https://wave.webaim.org/extension/)

#### Other

The system is definitely dependent on JavaScript, and possibly CSS, for full operation.

- [ ] Is useable without CSS
- [ ] Is useable without JS
- [x] Does not introduce new lint warnings
- [x] Flexible from small to large screens

### CODE REVIEW

File: cfgov/teachers_digital_platform/surveys.py
Comment (mrclay) [inline cfgov/teachers_digital_platform/surveys.py:48]: Use `@classmethod`.

File: cfgov/teachers_digital_platform/surveys.py
Comment (mrclay) [inline cfgov/teachers_digital_platform/surveys.py:53]: Use `@classmethod`.

File: cfgov/teachers_digital_platform/surveys.py
Comment (mrclay) [inline cfgov/teachers_digital_platform/surveys.py:249]: Use `@classmethod`.

**Review by mrclay — COMMENTED** [review]
Notes for self

File: cfgov/teachers_digital_platform/jinja2/teachers_digital_platform/survey/3-5/before-page-1.html
Comment (wpears) [inline cfgov/teachers_digital_platform/jinja2/teachers_digital_platform/survey/3-5/before-page-1.html:1]: Why are all these templates duplicated? The inclusion method would make sense if they were different, but since they're all the same, why not just bake this into `form-page.html`?

File: cfgov/teachers_digital_platform/jinja2/teachers_digital_platform/survey/results-3-5.html
Comment (wpears) [inline cfgov/teachers_digital_platform/jinja2/teachers_digital_platform/survey/results-3-5.html:1]: Similar to my comment on the before-page templates, why are all the result pages duplicated?

**Review by wpears — COMMENTED** [review]
Works as expected, aside from the yet-to-be-added WIP features.

One thing I'll note is that I think the radio buttons should use [our design system pattern](https://cfpb.github.io/design-system/components/radio-buttons). Particularly I think in the current iteration the list decoration + the radio button is messy/weird-looking.

Also, I'm a little squeamish about how once on the results page, the back button gets hijacked. I can imagine it would not be uncommon for users to want to step back through the pre-filled results without having to clear everything and start over. However, that doesn't seem possible in the current iteration.

It should also be noted that the share/print features are unusable if JS is disabled and as such should probably be only included on the page if JS is available.

File: cfgov/teachers_digital_platform/jinja2/teachers_digital_platform/survey/3-5/before-page-1.html
Comment (mrclay) [reply cfgov/teachers_digital_platform/jinja2/teachers_digital_platform/survey/3-5/before-page-1.html:1]: We expect them to differ once the content is updated, and, the include paths are generated programmatically so easier to have a little duplication than complex logic to reuse templates.

File: cfgov/teachers_digital_platform/jinja2/teachers_digital_platform/survey/results-3-5.html
Comment (mrclay) [reply cfgov/teachers_digital_platform/jinja2/teachers_digital_platform/survey/results-3-5.html:1]: We expect the content to differ. Once we analyze the full content we'll consolidate if we can.

**Comment by mrclay** [comment]
> One thing I'll note is that I think the radio buttons should use [our design system pattern](https://cfpb.github.io/design-system/components/radio-buttons).

Before launch I doubt we have remaining time to refactor--especially given the added difficulty of customizing Django's markup--AND get the team's approval, but I agree we should've nudged the design team towards this and we could do that after launch potentially.

> Also, I'm a little squeamish about how once on the results page, the back button gets hijacked. I can imagine it would not be uncommon for users to want to step back through the pre-filled results without having to clear everything and start over.

I'll ask the team if allowing backing into the survey (optionally readonly) would be OK. Currently the user _can_ just restart it.

> the share/print features are unusable if JS is disabled and as such should probably be only included on the page if JS is available.

Good idea.

CC @nbucknor @karenkitchens

**Comment by mrclay** [comment]
@wpears ...also big thanks for your time reviewing this!

**Review by wpears — COMMENTED** [review]
Looking really nice!

Visually, the only things I noticed are the obvious missing [LISTS] on the results page and it doesn't look like the bolding logic is working for the starting out / on the road / well on your way sections (starting out is always bolded, even if the text says something different).

Also it looks like we lost the button to go back and retake the survey, which will be especially needed if the back button is captured. Though as I said before, I'd much prefer if the routing rules were updated such that the back button wasn't captured. Moreover I think there's a bug in the routing, as if I finish a quiz for one grade level then manually return to the survey start page, when I try to take a different quiz I get sent to the results page for the first quiz.

**Comment by mrclay** [comment]
> if I finish a quiz for one grade level then manually return to the survey start page, when I try to take a different quiz I get sent to the results page for the first quiz.

This was intentional but in hindsight a strange decision. I'll ask the team about it.

**Comment by contolini** [comment]
Be careful about relying on client-side sanitization because a sneaky user can circumvent it and [manually encode some data](http://0.0.0.0:8000/consumer-tools/educator-tools/youth-financial-education/survey/view/?r=v1_6-8_h%3Ar%3Ah_3htn5%3AEpl4SQOkHa3P6oa6uBwExK4Dwvg#di41MnZ6ZmkuVHZzcz9mcGptP29+bXpxa2w/fnF7P2xren5zP2t3enZtP3xtent2az98fm17bD4+):

very important financial advice

**Comment by mrclay** [comment]
@contolini Thanks. I'll reject anything over 4 chars, but ultimately we don't want this PII even being sent to the servers.

**Comment by mrclay** [comment]
FYI we're working on later improvements in #6616. Please let us know if/when we should merge those in here.

**Review by wpears — APPROVED** [review]
Looks good!

Let's get https://github.com/cfpb/consumerfinance.gov/pull/6616 and https://github.com/cfpb/consumerfinance.gov/pull/6617 merged in here, then do one final pass before merging to master.

**Comment by mrclay** [comment]
@wpears Thanks, I've merged those and this is ready for release pending final review.

@higs4281 We moved python coverage ahead a bit here (and helped by removing some no-longer-needed code.

**Comment by wpears** [comment]
:+1:

**Pull request merged**
